memiavl/tree.go

```diff
@@ -121,3 +139,41 @@ func (t *Tree) Get(key []byte) []byte {
 func (t *Tree) Iterator(start, end []byte, ascending bool) dbm.Iterator {
 	return NewIterator(start, end, ascending, t.root)
 }
+
+func (t *Tree) ReplayWAL(untilVersion uint64, walPath string) error {
+	if untilVersion <= uint64(t.version) {
+		return fmt.Errorf("tree already up to date with untilVersion: %d with current version %d", untilVersion, t.version)
+	}
+
+	wal, err := wal.Open(walPath, nil)
+	if err != nil {
+		return fmt.Errorf("failed to open wal: %w", err)
+	}
+
+	var changesets []iavl.ChangeSet
+
+	// collect all changesets from WAL
+	for i := uint64(t.version + 1); i <= untilVersion; i++ {
+		bz, err := wal.Read(i)
+		if err != nil {
+			return fmt.Errorf("failed to read changeset with index %d from wal: %w", i, err)
+		}
+
+		blockChanges, err := UnmarshalChangeSet(bz)
+		if err != nil {
+			return err
+		}
+
+		changesets = append(changesets, *blockChanges)
```

Comment:
- I think we can call `ApplyChangeSet` here directly, so don't need to decode all the change sets in memory.
- we don't have to update hash for the intermediate versions, we can do `tree.ApplyChangeSet(, false)`, and only do a `tree.RootHash()` at the end.

Reply:
1) If we do that, if a user provides a version that does not exist in wal, the changes to the tree will be applied up to the version that exists in wal and then error. Is that desired?

Reply:
I also will add an option to simply sync all versions available in the wal

Reply:
added a way to sync untill the end of wal: if `untilVersion` is 0, it will sync till the end

Reply:
about the error still seems weird to me, it would mean that the tree is corrupted, but why would we declare it corrupted if that can be prevented? But the final decision is up to you

Reply:
also want to add function to read batches of changesets from wal in one go, will improve readability in `ReplayWAL` I believe

Reply:
nvm, I am using the native `wal.Read()` in `ReplayWAL`, not the wrapper, but I will propose batch reads to tidwall/wal

Reply:
> about the error still seems weird to me, it would mean that the tree is corrupted, but why would we declare it corrupted if that can be prevented? But the final decision is up to you

just want to lower the memory usage here, it could be pretty big depending on the snapshot interval.

Reply:
yeah your point makes sense. What do you think about copying the tree at the beginning of the function and applying changes to it the way you suggested? still a memory overhead, but not that big I guess

Reply:
I think if anything unexpected happens with WAL, it's almost always a serious unrecoverable issue.

Reply:
oh, I see what you mean. What I meant here is a user's mistake, aka providing an `untilVersion` that is higher than the latest version from the wal, which could result in `wal.Read()` error. I will just add a check that the version provided is within boundaries and we should be good to go with your design, thanks!